C++ OM: std::allocate_shared missing — a parse error that blocks uninstantiated templates

# C++ OM: `std::allocate_shared` missing — a parse error that blocks uninstantiated templates

`#6403` ("[om] Model shared_ptr, weak_ptr and make_shared") added a real
reference-counted `shared_ptr`, `weak_ptr` and `make_shared` to
`src/cpp/library/memory`. It did not add `std::allocate_shared`, and the way
that absence fails is worth its own issue: it is a **parse** error, not a link
error, so it takes out translation units that never call the function.

## Reproducer

No AWS headers, no flags:

```cpp
#include <memory>

/* Never instantiated -- naming allocate_shared is enough. */
template <typename T>
std::shared_ptr<T> MakeShared(const std::allocator<T> &alloc)
{
  return std::allocate_shared<T, std::allocator<T>>(alloc);
}

int main()
{
  return 0;
}
```

```
$ esbmc --std c++11 om_allocate_shared.cpp
om_allocate_shared.cpp:7:15: error: no member named 'allocate_shared' in namespace 'std'
    7 |   return std::allocate_shared<T, std::allocator<T>>(alloc);
      |               ^~~~~~~~~~~~~~~
om_allocate_shared.cpp:7:31: error: 'T' does not refer to a value
    7 |   return std::allocate_shared<T, std::allocator<T>>(alloc);
      |                               ^
ERROR: PARSING ERROR
```

`g++ -std=c++11 -fsyntax-only` accepts the same file.

Tested against `8d3cee251a` (8.4.0).

## Why the second diagnostic is the important one

With no declaration in scope, `std::allocate_shared` is not a template name, so
clang parses the `<` as less-than and the template argument `T` as a value.
That happens while **parsing the template definition**, before any
instantiation — so a function template that is never instantiated is enough to
fail the whole translation unit.

That is the difference between this and an ordinary missing-member gap. A
missing member that is only ever *called* fails at the call site, which a
harness can avoid. This one fails on inclusion, which nothing can avoid short of
a shim.

## Where it bites

`aws-cpp-sdk-core`. `Aws::MakeShared`
(`include/aws/core/utils/memory/stl/AWSAllocator.h:105-118`) is a function
template that Base64/hex codec verification never instantiates:

```cpp
template<typename T, typename ...ArgTypes>
std::shared_ptr<T> MakeShared(const char* allocationTag, ArgTypes&&... args)
{
    /* ... USE_AWS_MEMORY_MANAGEMENT bookkeeping elided ... */
    return std::allocate_shared<T, Aws::Allocator<T>>(Aws::Allocator<T>(), std::forward<ArgTypes>(args)...);
}
```

Merely including the header stops the run. This is the last item on the list
that started as #6183 — the `type_traits`/`unique_ptr`/`basic_string` gaps went
in #6190, the GOTO-conversion crash behind them in #6195, the
`basic_string(const char*, n)` false positive in #6225, and `shared_ptr` itself
in #6403. With `allocate_shared` declared, `aws-sdk-cpp` verification needs no
compatibility shim at all.

## Suggested fix, in rough order of cost

1. **Declare it.** Unblocks every translation unit that merely mentions it,
   which is the common case, and is a one-liner:

   ```cpp
   template <typename T, typename Alloc, typename... Args>
   shared_ptr<T> allocate_shared(const Alloc &, Args &&...);
   ```

2. **Define it in terms of the existing `make_shared`,** accepting that the
   allocator argument is ignored. Sound as long as the OM's `shared_ptr` is the
   thing doing the releasing, but it should carry a comment saying the allocator
   is not modelled — otherwise it silently answers questions about allocator
   behaviour that it has not modelled.

3. **Model allocator-aware allocation properly,** so a custom `Allocator`'s
   `allocate`/`deallocate` are actually called. Needed before anyone can verify
   a claim *about* an allocator, as opposed to code that merely uses one.

(1) is enough to unblock the reported case. (2) is what most users would expect
`allocate_shared` to do. Worth deciding deliberately rather than drifting into
(2) as an implementation detail of (1).

## Regression test

A `CHECK_FILE`-style test over the reproducer would catch the parse failure.
If (2) is taken, a test that the returned `shared_ptr` releases — a raw
allocation reported by `--memory-leak-check` as `forgotten memory`, versus the
same allocation through `allocate_shared` reporting `SUCCESSFUL` — would pin
the behaviour that matters. Note `--memory-leak-check` is meaningful against
this OM again now that `unique_ptr`'s destructor is no longer `#if 0`-ed out.

--- a/regression/esbmc-cpp11/cpp/allocate_shared_allocator_calls/main.cpp
+++ b/regression/esbmc-cpp11/cpp/allocate_shared_allocator_calls/main.cpp
@@ -1,0 +1,55 @@
+// github #6488, KNOWNBUG: pins the one thing the allocate_shared model gets
+// wrong. [util.smartptr.shared.create] requires the object's storage to come
+// from the supplied allocator, so g_allocs is 1 here -- g++ builds this and it
+// exits 0. ESBMC's <memory> routes allocate_shared through `new` instead
+// (shared_ptr releases with `delete`, so allocator-supplied malloc'd storage
+// would report a deallocation-operator mismatch), so allocate() is never
+// called, g_allocs stays 0, and ESBMC reports FAILED.
+//
+// This is the unsound direction: the same gap lets ESBMC *prove*
+// `g_allocs == 0`, a property false of every conforming implementation. Flip
+// this test to CORE when allocators are modelled for real.
+#include <memory>
+#include <cassert>
+#include <cstdlib>
+
+static int g_allocs = 0;
+
+template <typename T>
+struct CountingAlloc
+{
+  typedef T value_type;
+  CountingAlloc()
+  {
+  }
+  template <typename U>
+  CountingAlloc(const CountingAlloc<U> &)
+  {
+  }
+  T *allocate(size_t n)
+  {
+    ++g_allocs;
+    return (T *)malloc(n * sizeof(T));
+  }
+  void deallocate(T *p, size_t)
+  {
+    free(p);
+  }
+};
+
+struct S
+{
+  int x;
+  S(int v) : x(v)
+  {
+  }
+};
+
+int main()
+{
+  CountingAlloc<S> a;
+  std::shared_ptr<S> p = std::allocate_shared<S, CountingAlloc<S>>(a, 7);
+  assert(p->x == 7);
+  assert(g_allocs == 1);
+  return 0;
+}

--- a/regression/esbmc-cpp11/cpp/allocate_shared_allocator_calls/test.desc
+++ b/regression/esbmc-cpp11/cpp/allocate_shared_allocator_calls/test.desc
@@ -1,0 +1,4 @@
+KNOWNBUG
+main.cpp
+--std c++11
+^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp11/cpp/allocate_shared_basic/main.cpp
+++ b/regression/esbmc-cpp11/cpp/allocate_shared_basic/main.cpp
@@ -1,0 +1,38 @@
+// github #6488: allocate_shared forwards its arguments to T's constructor and
+// hands the result to the same reference-counted shared_ptr make_shared uses,
+// so the object is released when the last owner dies. The allocator argument
+// is accepted but not modelled.
+#include <memory>
+#include <cassert>
+
+struct S
+{
+  int x;
+  int y;
+  S(int a, int b) : x(a), y(b)
+  {
+  }
+};
+
+int main()
+{
+  std::allocator<S> alloc;
+
+  std::shared_ptr<S> p = std::allocate_shared<S, std::allocator<S>>(alloc, 3, 4);
+  assert(p->x == 3);
+  assert(p->y == 4);
+  assert(p.use_count() == 1);
+
+  {
+    std::shared_ptr<S> q = p;
+    assert(q.use_count() == 2);
+    assert(q.get() == p.get());
+  }
+  assert(p.use_count() == 1); // the copy released its share
+
+  std::shared_ptr<int> n =
+    std::allocate_shared<int, std::allocator<int>>(std::allocator<int>());
+  assert(*n == 0);
+
+  return 0;
+}

--- a/regression/esbmc-cpp11/cpp/allocate_shared_basic/main.cpp
+++ b/regression/esbmc-cpp11/cpp/allocate_shared_basic/main.cpp
@@ -1,9 +1,10 @@
 // github #6488: allocate_shared forwards its arguments to T's constructor and
 // hands the result to the same reference-counted shared_ptr make_shared uses,
 // so the object is released when the last owner dies. The allocator argument
-// is accepted but not modelled.
+// is accepted but not modelled -- see allocate_shared_allocator_calls.
 #include <memory>
 #include <cassert>
+#include <cstdlib>
 
 struct S
 {
@@ -11,6 +12,29 @@ struct S
   int y;
   S(int a, int b) : x(a), y(b)
   {
+  }
+};
+
+// The reason allocate_shared exists: a user-defined allocator, which is what
+// real callers pass, with A deduced rather than spelled out.
+template <typename T>
+struct MyAlloc
+{
+  typedef T value_type;
+  MyAlloc()
+  {
+  }
+  template <typename U>
+  MyAlloc(const MyAlloc<U> &)
+  {
+  }
+  T *allocate(size_t n)
+  {
+    return (T *)malloc(n * sizeof(T));
+  }
+  void deallocate(T *p, size_t)
+  {
+    free(p);
   }
 };
 
@@ -29,6 +53,11 @@ int main()
     assert(q.get() == p.get());
   }
   assert(p.use_count() == 1); // the copy released its share
+
+  std::shared_ptr<S> d = std::allocate_shared<S>(MyAlloc<S>(), 5, 6);
+  assert(d->x == 5);
+  assert(d->y == 6);
+  assert(d.use_count() == 1);
 
   std::shared_ptr<int> n =
     std::allocate_shared<int, std::allocator<int>>(std::allocator<int>());

--- a/regression/esbmc-cpp11/cpp/allocate_shared_basic/test.desc
+++ b/regression/esbmc-cpp11/cpp/allocate_shared_basic/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.cpp
+--std c++11 --memory-leak-check
+^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp11/cpp/allocate_shared_fail/main.cpp
+++ b/regression/esbmc-cpp11/cpp/allocate_shared_fail/main.cpp
@@ -1,0 +1,21 @@
+// github #6488: the negative side of allocate_shared_basic -- the constructor
+// arguments really are forwarded, so a claim about the wrong value is refuted
+// rather than vacuously proved.
+#include <memory>
+#include <cassert>
+
+struct S
+{
+  int x;
+  S(int v) : x(v)
+  {
+  }
+};
+
+int main()
+{
+  std::shared_ptr<S> p =
+    std::allocate_shared<S, std::allocator<S>>(std::allocator<S>(), 3);
+  assert(p->x == 4);
+  return 0;
+}

--- a/regression/esbmc-cpp11/cpp/allocate_shared_fail/test.desc
+++ b/regression/esbmc-cpp11/cpp/allocate_shared_fail/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.cpp
+--std c++11
+^VERIFICATION FAILED$

--- a/regression/esbmc-cpp11/cpp/allocate_shared_fail/test.desc
+++ b/regression/esbmc-cpp11/cpp/allocate_shared_fail/test.desc
@@ -2,3 +2,4 @@ CORE
 main.cpp
 --std c++11
 ^VERIFICATION FAILED$
+assertion p->x == 4

--- a/regression/esbmc-cpp11/cpp/allocate_shared_uninstantiated/main.cpp
+++ b/regression/esbmc-cpp11/cpp/allocate_shared_uninstantiated/main.cpp
@@ -1,0 +1,16 @@
+// github #6488: <memory> declared make_shared but not allocate_shared, so
+// merely *naming* it made clang parse the `<` as less-than and reject the
+// translation unit -- even from a function template that is never
+// instantiated, which no harness can work around.
+#include <memory>
+
+template <typename T>
+std::shared_ptr<T> MakeShared(const std::allocator<T> &alloc)
+{
+  return std::allocate_shared<T, std::allocator<T>>(alloc);
+}
+
+int main()
+{
+  return 0;
+}

--- a/regression/esbmc-cpp11/cpp/allocate_shared_uninstantiated/test.desc
+++ b/regression/esbmc-cpp11/cpp/allocate_shared_uninstantiated/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.cpp
+--std c++11
+^VERIFICATION SUCCESSFUL$

--- a/src/cpp/library/memory
+++ b/src/cpp/library/memory
@@ -867,6 +867,18 @@ shared_ptr<T> make_shared(Args &&...args)
 {
   return shared_ptr<T>(new T(std::forward<Args>(args)...));
 }
+
+/* The allocator is deliberately not modelled: storage comes from the same
+ * `new` make_shared uses, because shared_ptr's release path is `delete __p`.
+ * So a custom allocator's allocate/deallocate/construct are never called, and
+ * this model cannot answer questions *about* an allocator -- only about code
+ * that happens to use one (github #6488). The array forms added in C++20 are
+ * not declared. */
+template <class T, class A, class... Args>
+shared_ptr<T> allocate_shared(const A &, Args &&...args)
+{
+  return make_shared<T>(std::forward<Args>(args)...);
+}
 #endif
 
 #if 0

--- a/src/cpp/library/memory
+++ b/src/cpp/library/memory
@@ -868,13 +868,19 @@ shared_ptr<T> make_shared(Args &&...args)
   return shared_ptr<T>(new T(std::forward<Args>(args)...));
 }
 
-/* The allocator is deliberately not modelled: storage comes from the same
- * `new` make_shared uses, because shared_ptr's release path is `delete __p`.
- * So a custom allocator's allocate/deallocate/construct are never called, and
- * this model cannot answer questions *about* an allocator -- only about code
- * that happens to use one (github #6488). The array forms added in C++20 are
- * not declared. */
-template <class T, class A, class... Args>
+/* The allocator is deliberately not modelled, and the two halves cannot be
+ * coupled as they stand: allocator::allocate hands out malloc'd storage while
+ * shared_ptr releases with `delete __p`, so routing through the allocator
+ * makes every use report "Mismatched memory deallocation operators". Storage
+ * therefore comes from the same `new` make_shared uses, a custom allocator's
+ * allocate/deallocate/construct are never called, and this model cannot
+ * answer questions *about* an allocator -- only about code that happens to
+ * use one. allocate_shared_allocator_calls pins that divergence (github
+ * #6488). The array forms added in C++20 are not declared.
+ * `typename A::value_type` ([allocator.requirements]) stops A deducing to a
+ * constructor argument when the allocator is omitted altogether, which would
+ * otherwise silently swallow it. */
+template <class T, class A, class... Args, class = typename A::value_type>
 shared_ptr<T> allocate_shared(const A &, Args &&...args)
 {
   return make_shared<T>(std::forward<Args>(args)...);

